### PR TITLE
[X86] For minsize memset/memcpy, use byte or double-word accesses

### PR DESCRIPTION
**Pull request by AZero13**
repstosb and repstosd are the same size, but stosd is only done for 0 because the process of multiplying the constant so that it is copied across the bytes of the 32-bit number adds extra instructions that cause the size to increase. For 0, repstosb and repstosd are the same size, but stosd is only done for 0 because the process of multiplying the constant so that it is copied across the bytes of the 32-bit number adds extra instructions that cause the size to increase. For 0, we do not need to do that at all.

For memcpy, the same goes, and as a result the minsize check was moved ahead because a jmp to memcpy encoded takes more bytes than repmovsb.

### CODE REVIEW

File: llvm/lib/Target/X86/X86SelectionDAGInfo.cpp [hunk 1]
```diff
@@ -67,10 +67,27 @@ SDValue X86SelectionDAGInfo::EmitTargetCodeForMemset(
...
 
+  // If we have minsize, then don't care about the alignment.
+  // On x86, the CPU doesn't care and neither should you.
+  // As long as the count is aligned, we can use the minimum number of
+  // instructions without always having to resort to strosb.
+  //
+  // Because this is a feature specific to x86, we must handle it here.
   uint64_t SizeVal = ConstantSize->getZExtValue();
+  if (DAG.getMachineFunction().getFunction().hasMinSize()) {
+    if ((SizeVal & 7) == 0 && Subtarget.is64Bit())
```
Comment (topperc) [inline llvm/lib/Target/X86/X86SelectionDAGInfo.cpp:81]: I think you can do something like `commonAlignment(Align(Subtarget.is64Bit() ? 8 : 4), SizeVal)` instead of this if/else chain.

File: llvm/lib/Target/X86/X86SelectionDAGInfo.cpp [hunk 2]
```diff
@@ -67,10 +67,27 @@ SDValue X86SelectionDAGInfo::EmitTargetCodeForMemset(
...
   // address value and run time information about the CPU.
   if (Alignment < Align(4) || !ConstantSize ||
-      ConstantSize->getZExtValue() > Subtarget.getMaxInlineSizeThreshold()) 
+      ConstantSize->getZExtValue() > Subtarget.getMaxInlineSizeThreshold())
     return SDValue();
 
+  // If we have minsize, then don't care about the alignment.
+  // On x86, the CPU doesn't care and neither should you.
+  // As long as the count is aligned, we can use the minimum number of
+  // instructions without always having to resort to strosb.
```
Comment (topperc) [inline llvm/lib/Target/X86/X86SelectionDAGInfo.cpp:76]: strosb -> stosb

**Comment by topperc** [comment]
Title should mention "memset". Otherwise its unclear what case you're talking about.

File: llvm/lib/Target/X86/X86SelectionDAGInfo.cpp [hunk 3]
```diff
@@ -66,11 +66,30 @@ SDValue X86SelectionDAGInfo::EmitTargetCodeForMemset(
...
+        ConstantSize->getZExtValue() > Subtarget.getMaxInlineSizeThreshold())))
     return SDValue();
 
+  // If we have minsize, then don't care about the alignment.
+  // On x86, the CPU doesn't care and neither should you.
+  // As long as the count is aligned, we can use the minimum number of
+  // instructions without always having to resort to stosb.
+  //
+  // Because this is a feature specific to x86, we must handle it here.
   uint64_t SizeVal = ConstantSize->getZExtValue();
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86SelectionDAGInfo.cpp:81]: Move this up and reuse?

Comment (RKSimon) [reply llvm/lib/Target/X86/X86SelectionDAGInfo.cpp:81]: (See #87825 - split the if statement to simplify)

**Comment by AtariDreams** [comment]
Ping?

**Review by RKSimon — COMMENTED** [review]
Do we have a test case with the STOSB prologue stores?

**Comment by AtariDreams** [comment]
@RKSimon This should be good now!

**Comment by topperc** [comment]
As I wrote previously, the title should mention memset

**Comment by AtariDreams** [comment]
> As I wrote previously, the title should mention memset

Done!

**Comment by AtariDreams** [comment]
> Do we have a test case with the STOSB prologue stores?

Yes

File: llvm/test/CodeGen/X86/memset-vs-memset-inline.ll [hunk 4]
```diff
@@ -28,138 +28,22 @@ define void @regular_memset_calls_external_function(ptr %a, i8 %value) nounwind
...
-; CHECK-NEXT:    movq %rax, 64(%rdi)
-; CHECK-NEXT:    movq %rax, 56(%rdi)
-; CHECK-NEXT:    movq %rax, 48(%rdi)
-; CHECK-NEXT:    movq %rax, 40(%rdi)
-; CHECK-NEXT:    movq %rax, 32(%rdi)
-; CHECK-NEXT:    movq %rax, 24(%rdi)
-; CHECK-NEXT:    movq %rax, 16(%rdi)
-; CHECK-NEXT:    movq %rax, 8(%rdi)
-; CHECK-NEXT:    movq %rax, (%rdi)
+; CHECK-NEXT:    movl %esi, %eax
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/memset-vs-memset-inline.ll:31]: Why this changed without minsize?

Comment (goldsteinn) [reply llvm/test/CodeGen/X86/memset-vs-memset-inline.ll:31]: I think the:
```

  if (!AlwaysInline &&
      (Alignment < Align(4) || SizeVal > Subtarget.getMaxInlineSizeThreshold()))
```

With `AlwaysInline` being `true` here we fail on the prior check.

File: llvm/lib/Target/X86/X86SelectionDAGInfo.cpp [hunk 5]
```diff
@@ -62,18 +62,31 @@ SDValue X86SelectionDAGInfo::EmitTargetCodeForMemset(
   const X86Subtarget &Subtarget =
       DAG.getMachineFunction().getSubtarget<X86Subtarget>();
 
+  // If we have minsize, then don't care about the alignment.
+  // On x86, the CPU doesn't care and neither should you.
+  // As long as the count is aligned, we can use the minimum number of
+  // instructions without always having to resort to stosb.
+  //
+  // Because this is a feature specific to x86, we must handle it here.
```
Comment (goldsteinn) [inline llvm/lib/Target/X86/X86SelectionDAGInfo.cpp:70]: Why do you need count aligned? Just use `rep stosb` (likewise with minsize we shouldn't try to promote `rep stosb` to 16/32/64 bit if we have minsize (and probably not if we have `ERMS` in general but that should be an independent patch)).

Comment (AtariDreams) [reply llvm/lib/Target/X86/X86SelectionDAGInfo.cpp:70]: are these instructions different sizes when encoded? if not, then I don't see why not promote rep stosb

Comment (goldsteinn) [reply llvm/lib/Target/X86/X86SelectionDAGInfo.cpp:70]: Yes.
```
       0: f3 aa                        	rep		stosb	%al, %es:(%rdi)
       2: f3 66 ab                     	rep		stosw	%ax, %es:(%rdi)
       5: f3 ab                        	rep		stosd	%eax, %es:(%rdi)
       7: f3 48 ab                     	rep		stosq	%rax, %es:(%rdi)
```
Further if you have minsize to shrink the cost of encoding the `imm` (especially problematic for `imm64`). Further things the ERMS only really require that the byte-variant be fast (although every processor I'm aware of with ERMS also has fast 2/4/8 byte variants).

File: llvm/lib/Target/X86/X86SelectionDAGInfo.cpp [hunk 6]
```diff
@@ -146,8 +154,8 @@ SDValue X86SelectionDAGInfo::EmitTargetCodeForMemset(
                     DAG.getNode(ISD::ADD, dl, AddrVT, Dst,
                                 DAG.getConstant(Offset, dl, AddrVT)),
                     Val, DAG.getConstant(BytesLeft, dl, SizeVT), Alignment,
-                    isVolatile, AlwaysInline,
-                    /* CI */ nullptr, DstPtrInfo.getWithOffset(Offset)));
+                    isVolatile, /*AlwaysInline*/ true, /* CI */ nullptr,
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86SelectionDAGInfo.cpp:157]: Why have you overriden AlwaysInline to true?

Comment (RSilicon) [reply llvm/lib/Target/X86/X86SelectionDAGInfo.cpp]: Not anymore.

Comment (RSilicon) [reply llvm/lib/Target/X86/X86SelectionDAGInfo.cpp]: Though I did this to match memcpy, which does do this, and I don't think we should do a call when we know that this is for 7 bytes at most.

Comment (RSilicon) [reply llvm/lib/Target/X86/X86SelectionDAGInfo.cpp]: Anyway, I undid that already.

**Comment by RSilicon** [comment]
@RKSimon I believe this is ready now!

**Comment by RSilicon** [comment]
@phoebewang I believe this is ready!

**Review by RKSimon — COMMENTED** [review]
Your patch describes memset yet most of the changes appear to be for memcpy :/

**Comment by RSilicon** [comment]
> Your patch describes memset yet most of the changes appear to be for memcpy :/

Incorrect. The memset code was refactored to use the same method as the memcpy code for determining block size.

**Comment by topperc** [comment]
> > Your patch describes memset yet most of the changes appear to be for memcpy :/
> 
> Incorrect. The memset code was refactored to use the same method as the memcpy code for determining block size.

You changed the codegen for 2 memcpy tests, but do not mention that in the patch description.

**Comment by RSilicon** [comment]
> > > Your patch describes memset yet most of the changes appear to be for memcpy :/
> > 
> > 
> > Incorrect. The memset code was refactored to use the same method as the memcpy code for determining block size.
> 
> You changed the codegen for 2 memcpy tests, but do not mention that in the patch description.

Done!

**Comment by RSilicon** [comment]
@topperc Is this adequate?

**Comment by RSilicon** [comment]
@phoebewang is this adequate?

**Comment by phoebewang** [comment]
> @phoebewang is this adequate?

Can you split the 2nd commit into one NFC for refactor and one for the change?

**Comment by AreaZR** [comment]
> repstosb and repstosd are the same size, but stosd is only done for 0 because the process of multiplying the constant so that it is copied across the bytes of the 32-bit number adds extra instructions that cause the size to increase. For 0, repstosb and repstosd are the same size, but stosd is only done for 0 because the process of multiplying the constant so that it is copied across the bytes of the 32-bit number adds extra instructions that cause the size to increase. For 0, we do not need to do that at all.
> 
> For memcpy, the same goes, and as a result the minsize check was moved ahead because a jmp to memcpy encoded takes more bytes than repmovsb.

Done!

**Review by phoebewang — APPROVED** [review]
LGTM.

**Comment by AreaZR** [comment]
Thank you @phoebewang. Can we please merge?

**Comment by zmodem** [comment]
We're hitting an assert after this change:

```
$ cat /tmp/bug.ll
target triple = "i686-pc-windows-msvc"

declare void @llvm.memset.p0.i32(ptr nocapture writeonly, i8, i32, i1 immarg)

define dso_local void @foo(ptr %0) {
  call void @llvm.memset.p0.i32(ptr noundef nonnull align 4 dereferenceable(90) %0, i8 -1, i32 90, i1 false)
  ret void
}

$ build/bin/llc /tmp/bug.ll
llc: /work/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:7736: SDValue getMemsetValue(SDValue, EVT, SelectionDAG &, const SDLoc &): Assertion `C->getAPIntValue().getBitWidth() == 8' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: build/bin/llc /tmp/bug.ll
1.      Running pass 'Function Pass Manager' on module '/tmp/bug.ll'.
2.      Running pass 'X86 DAG->DAG Instruction Selection' on function '@foo'
 #0 0x00005557413cadb8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build/bin/llc+0x367adb8)
 #1 0x00005557413c896e llvm::sys::RunSignalHandlers() (build/bin/llc+0x367896e)
 #2 0x00005557413cb448 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f5003c591a0 (/lib/x86_64-linux-gnu/libc.so.6+0x3d1a0)
 #4 0x00007f5003ca70ec __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007f5003c59102 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #6 0x00007f5003c424f2 abort ./stdlib/abort.c:81:7
 #7 0x00007f5003c42415 _nl_load_domain ./intl/loadmsgcat.c:1177:9
 #8 0x00007f5003c51d32 (/lib/x86_64-linux-gnu/libc.so.6+0x35d32)
 #9 0x00005557411b5e69 getMemsetValue(llvm::SDValue, llvm::EVT, llvm::SelectionDAG&, llvm::SDLoc const&) SelectionDAG.cpp:0:0
#10 0x0000555741193faa getMemsetStores(llvm::SelectionDAG&, llvm::SDLoc const&, llvm::SDValue, llvm::SDValue, llvm::SDValue, unsigned long, llvm::Align, bool, bool, llvm::MachinePointerInfo, llvm::AAMDNodes const&) SelectionDAG.cpp:0:0
#11 0x0000555741192f43 llvm::SelectionDAG::getMemset(llvm::SDValue, llvm::SDLoc const&, llvm::SDValue, llvm::SDValue, llvm::SDValue, llvm::Align, bool, bool, llvm::CallInst const*, llvm::MachinePointerInfo, llvm::AAMDNodes const&) (build/bin/llc+0x3442f43)
#12 0x000055573fe2fff6 llvm::X86SelectionDAGInfo::EmitTargetCodeForMemset(llvm::SelectionDAG&, llvm::SDLoc const&, llvm::SDValue, llvm::SDValue, llvm::SDValue, llvm::SDValue, llvm::Align, bool, bool, llvm::MachinePointerInfo) const (build/bin/llc+0x20dfff6)
#13 0x000055574119301d llvm::SelectionDAG::getMemset(llvm::SDValue, llvm::SDLoc const&, llvm::SDValue, llvm::SDValue, llvm::SDValue, llvm::Align, bool, bool, llvm::CallInst const*, llvm::MachinePointerInfo, llvm::AAMDNodes const&) (build/bin/llc+0x344301d)
#14 0x000055574113bb51 llvm::SelectionDAGBuilder::visitIntrinsicCall(llvm::CallInst const&, unsigned int) (build/bin/llc+0x33ebb51)
#15 0x0000555741107770 llvm::SelectionDAGBuilder::visit(llvm::Instruction const&) (build/bin/llc+0x33b7770)
#16 0x00005557411c7715 llvm::SelectionDAGISel::SelectBasicBlock(llvm::ilist_iterator_w_bits<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void, true, llvm::BasicBlock>, false, true>, llvm::ilist_iterator_w_bits<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void, true, llvm::BasicBlock>, false, true>, bool&) (build/bin/llc+0x3477715)
#17 0x00005557411c69cd llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (build/bin/llc+0x34769cd)
#18 0x00005557411c3e41 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (build/bin/llc+0x3473e41)
#19 0x00005557411c1706 llvm::SelectionDAGISelLegacy::runOnMachineFunction(llvm::MachineFunction&) (build/bin/llc+0x3471706)
#20 0x0000555740557da6 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (build/bin/llc+0x2807da6)
#21 0x0000555740a9ee69 llvm::FPPassManager::runOnFunction(llvm::Function&) (build/bin/llc+0x2d4ee69)
#22 0x0000555740aa7362 llvm::FPPassManager::runOnModule(llvm::Module&) (build/bin/llc+0x2d57362)
#23 0x0000555740a9f93c llvm::legacy::PassManagerImpl::run(llvm::Module&) (build/bin/llc+0x2d4f93c)
#24 0x000055573e8901c7 main (build/bin/llc+0xb401c7)
#25 0x00007f5003c43b8a __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#26 0x00007f5003c43c45 call_init ./csu/../csu/libc-start.c:128:20
#27 0x00007f5003c43c45 __libc_start_main ./csu/../csu/libc-start.c:347:5
#28 0x000055573e889ec1 _start (build/bin/llc+0xb39ec1)
Aborted
```

I'll revert until this gets fixed.

**Comment by AreaZR** [comment]
@zmodem Addressed! https://github.com/llvm/llvm-project/pull/111393
